strawberry/relay/types.py

```diff
@@ -0,0 +1,706 @@
+import abc
+import dataclasses
+import inspect
+import itertools
+import sys
+import uuid
+from typing import (
+    Any,
+    Awaitable,
+    ClassVar,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Optional,
+    Sequence,
+    Sized,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+)
+from typing_extensions import Literal, get_args, get_origin
+
+from strawberry.field import field
+from strawberry.lazy_type import LazyType
+from strawberry.object_type import interface, type
+from strawberry.scalars import ID
+from strawberry.type import StrawberryContainer
+from strawberry.types.info import Info
+from strawberry.types.types import TypeDefinition
+from strawberry.utils.await_maybe import AwaitableOrValue
+
+from .utils import from_base64, to_base64
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+connection_typename = "arrayconnection"
```

Comment:
could we avoid to use this somehow? I think we can always use the type name from strawberry, no? :)

Reply:
This actually is the name of the cursor (maybe it is poorly named? =P)

It is used in here: https://github.com/strawberry-graphql/strawberry/pull/2511/files#diff-87ff2a77ed8a7dc2eb7fc8a320e55d457b77160ab7932c185b8becbf8662767bR526 . I used the same name that [graphql-relay-py](https://github.com/graphql-python/graphql-relay-py/blob/97ba15e573eabcefd13e249d05fc5557f0296804/src/graphql_relay/connection/array_connection.py#L158) uses.

I probably wanted to call this `cursor_typename` but wrote `connection_typename` by mistake. Do you have any suggestions for a better name?

Reply:
oh I mean `arrayconnection`, wouldn't be best to use the name of the types we are going to return? we do have access to those, no?

Reply:
They are actually not used in type names, but in the cursor. For example:

```json
{
  "edges" [
    {
      "cursor": "arrayconnection:0",  # this
      "node": {
        "__typename": "Fruit",
        "name": "Banana"
      },
    },
    {
      "cursor": "arrayconnection:1",  # this
      "node": {
        "__typename": "Fruit",
        "name": "Apple"
      },
    },
  ]
}
```

It is used in those places, but it is exposed as `base64` (I put them raw in the example above to show where they are used).

The cursor has the pagination information, which is not necessarily related to the type itself. For the base implementation, which uses limit/offset, it will be the row number of the result. So, if your connection has a total number objects of `1000`, you would have cursors from `arrayconnection:0` to `arrayconnection:999`. Then when asking for `first: 10` and `after: arrayconnection:50` (but in base64) for example your results would have edges from `arrayconnection:51` to `arrayconnection:61`

So the first cursor in a connection of 2 different types will both be `arrayconnection:0` because they are both the rownumber 0 in that connection.

Regarding the name, it is also used in this example (https://relay.dev/docs/guides/graphql-server-specification/). If you decode the base64 inside the cursor, you can see that it also uses `arrayconnection` for it